#Exercise 1

In [1]:
!pip install -q pyomo

In [2]:
from pyomo.environ import *

In [3]:
import numpy as np

In [4]:
model = ConcreteModel()

In [5]:
N = 3
M = 3
lb = np.array([1,2,1])
ub = np.array([np.inf,np.inf,np.inf])
obj_coeff = np.array([1,-2,3])
constr_coeff_A = np.array([[2,1,2],[-1,1,2],[2,-3,1]])
constr_rhs_b = np.array([10,-3,-6])
row_indices = np.arange(M)
col_indices = np.arange(N)

In [6]:
model.x = Var(col_indices)
model.constraints = ConstraintList()

In [7]:
for i in col_indices:
  model.x[i].setlb(lb[i])
  model.x[i].setub(ub[i])

In [8]:
for i in row_indices:
  model.constraints.add(sum(constr_coeff_A[i][j]*model.x[j] for j in col_indices) <= constr_rhs_b[i])

In [9]:
model.objective = Objective(expr = summation(obj_coeff,model.x), sense = minimize)
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     1 :  None :   inf : False :  True :  Reals
          1 :     2 :  None :   inf : False :  True :  Reals
          2 :     1 :  None :   inf : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : x[0] - 2.0*x[1] + 3*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                   : Upper : Active
          1 :  -Inf : 2*x[0] + x[1] + 2*x[2] :  10.0 

In [10]:
!apt-get install -y -qq glpk-utils

In [11]:
opt_glpk = SolverFactory('glpk', executable='/usr/bin/glpsol')
result_glpk = opt_glpk.solve(model)
print(result_glpk)
print('Solver status:', result_glpk.solver.status)
print('Solver termination condition:',result_glpk.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 10
  Sense: minimize
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011620521545410156

Solver status: ok
Solver termination condition: other


In [12]:
!apt-get install -y -qq coinor-cbc

In [13]:
opt_cbc = SolverFactory('cbc')
result_cbc = opt_cbc.solve(model)
print(result_cbc)
print('Solver status:', result_cbc.solver.status)
print('Solver termination condition:',result_cbc.solver.termination_condition)

    model.name="unknown";
      - termination condition: infeasible
      - message from solver: <undefined>

Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.016359806060791016

Solver status: warning
Solver termination condition: infeasible


## Ans 1 :

When we use glpk solver, the solver status is ok and solver termination condition is other.
It is not meaningful as it does not tell what is actually the termination condition.

Whereas when we use cbc solver, the solver status is warning and solver termination condition is infeasible.
It has more meaning as we get to know that our linear program is infeasible and it also gives us a warning relating to the conflict in the constraints.

Possible reasons for these messages could be that the constraints are conflicting.


In [14]:
model.objective.set_sense(maximize)

In [15]:
result_cbc = opt_cbc.solve(model)
print(result_cbc)
print('Solver status:', result_cbc.solver.status)
print('Solver termination condition:',result_cbc.solver.termination_condition)

    model.name="unknown";
      - termination condition: infeasible
      - message from solver: <undefined>

Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: None
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: maximize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.02814197540283203

Solver status: warning
Solver termination condition: infeasible


##Ans 2(a) :

Solver Status is warning.
Solver termination condition is infeasible.

There does not exist any solution that satisfies all the given constraints and hence it is infeasible.

In [16]:
model.objective.set_sense(minimize)

In [17]:
model.x[1].setub(8)

In [18]:
result_cbc = opt_cbc.solve(model)
print(result_cbc)
print('Solver status:', result_cbc.solver.status)
print('Solver termination condition:',result_cbc.solver.termination_condition)

    model.name="unknown";
      - termination condition: infeasible
      - message from solver: <undefined>

Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.02325129508972168

Solver status: warning
Solver termination condition: infeasible


##Ans 2(b) :

Solver Status is warning.
Solver termination condition is infeasible.
There does not exist any solution that satisfies all the given constraints and hence it is infeasible.

In [19]:
model.x[1].setub(np.inf)

In [20]:
new_constr_coeff = [0,-1,1]
new_constr_coeff_b = 9
model.constraints.add(sum(new_constr_coeff[i]*model.x[i] for i in col_indices) <= new_constr_coeff_b)

In [21]:
result_cbc = opt_cbc.solve(model)
print(result_cbc)
print('Solver status:', result_cbc.solver.status)
print('Solver termination condition:',result_cbc.solver.termination_condition)

    model.name="unknown";
      - termination condition: infeasible
      - message from solver: <undefined>

Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.019324064254760742

Solver status: warning
Solver termination condition: infeasible


##Ans 2(c) :

Solver Status is warning.
Solver termination condition is infeasible.
There does not exist any solution that satisfies all the given constraints and hence it is infeasible.

In [22]:
model.constraints[4].deactivate()

In [23]:
model.objective.deactivate()

In [24]:
new_obj_coeff = np.array([1,1,1])
model.new_objective = Objective(expr = summation(new_obj_coeff,model.x), sense = minimize)
model.constraints[1].deactivate()
new_constr_coeff_1 = [1,1,0]
new_constr_coeff_b_1 = 45
model.constraints.add(sum(new_constr_coeff_1[i]*model.x[i] for i in col_indices) >= 45)

In [25]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     1 :  None :   inf : False :  True :  Reals
          1 :     2 :  None :   inf : False :  True :  Reals
          2 :     1 :  None :   inf : False :  True :  Reals

2 Objective Declarations
    new_objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : x[0] + x[1] + x[2]
    objective : Size=1, Index=None, Active=False
        Key  : Active : Sense    : Expression
        None :  False : minimize : x[0] - 2.0*x[1] + 3*x[2]

1 Constraint Declarations
    constraints : S

In [26]:
result_cbc = opt_cbc.solve(model)
print(result_cbc)
print('Solver status:', result_cbc.solver.status)
print('Solver termination condition:',result_cbc.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 46.0
  Upper bound: 46.0
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 2
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 2
  Error rc: 0
  Time: 0.018092870712280273
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal


In [27]:
print('Objective = ', model.new_objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)
print('\nConstraints')
model.constraints.display()

Objective =  46.0

Decision Variables
x[ 0 ] =  25.6
x[ 1 ] =  19.4
x[ 2 ] =  1.0

Constraints
constraints : Size=5
    Key : Lower : Body               : Upper
      2 :  None : -4.200000000000003 :  -3.0
      3 :  None : -5.999999999999993 :  -6.0
      5 :  45.0 :               45.0 :  None


##Ans 2(d) :

The value of the objective function is : 46.0

The values of decision variables are:

$x_1 = 25.6\\x_2 = 19.4\\x_3 = 1.0$

The constraint 1 : $-x_1+x_2+2x_3\leq-3$ is not active as it is not satisifed to equality.

The constraint 2 : $2x_1-3x_2+x_3\leq-6$ is not active as it is not satisifed to equality.

The constraint 3 : $x_1+x_2\geq45$ is active as it is satisifed to equality.